java/arcs/core/storage/CapabilitiesResolver.kt

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.storage
+
+import arcs.core.common.ArcId
+import arcs.core.data.Capabilities
+import arcs.core.util.TaggedLog
+
+/**
+ * [CapabilitiesResolver] is a factory class that creates [StorageKey]s based on the registered
+ * providers and given [Capabilities].
+ */
+class CapabilitiesResolver(
+    val options: StorageKeyOptions,
+    private val creators: StorageKeyCreatorsMap = getAllCreators()
+) {
+    private val log = TaggedLog { "CapabilitiesResolver" }
+
+    /* Options used to construct [CapabilitiesResolver] */
+    data class StorageKeyOptions(val arcId: ArcId)
+
+    /* Creates and returns a [StorageKey] corresponding to the given [Capabilities]. */
+    fun createStorageKey(
+        capabilities: Capabilities,
+        entitySchemaHash: String? = null
+    ): StorageKey? {
+        // TODO: This is a naive and basic solution for picking the appropriate
+        // storage key creator for the given capabilities. As more capabilities are
+        // added the heuristics will become more robust.
+        val protocols = findStorageKeyProtocols(capabilities)
+        if (protocols.isEmpty()) {
+            throw IllegalStateException(
+                "Cannot create a suitable storage key for $capabilities"
+            )
+        } else if (protocols.size > 1) {
+            log.warning { "Multiple storage key creators for $capabilities" }
+        }
+        return creators[protocols.first()]?.second?.invoke(
+            options,
+            entitySchemaHash ?: ""
+        )
+    }
+
+    /* Returns set of protocols corresponding to the given [Capabilities]. */
+    /* internal */ fun findStorageKeyProtocols(capabilities: Capabilities): Set<String> {
+        val protocols: MutableSet<String> = mutableSetOf()
+        for ((protocol, creator) in creators) {
+            if (creator.first.contains(capabilities)) {
+                protocols.add(protocol)
+            }
+        }
```

Comment:
Optional, prefer using kotlin collection iterators like so:

```kotlin
creators.forEach { (protocol, creator) -> 
    if (capabilities in creator.first) protocols.add(protocol)
}
```

Also notice the use of `in` instead of `.contains`

Reply:
Ah right! sorry - i forgot that creator.first doesn't support `in`.

Though - it could be added somewhat easily using `override operator fun contains(...): Boolean`

Reply:
made an operator and using `in` now.
`creators` are a map and internally checks aren't happy with using `forEach` on maps.